bundles/org.openhab.binding.withings/src/main/java/org/openhab/binding/withings/internal/handler/WithingsBridgeHandler.java

```diff
@@ -0,0 +1,239 @@
+    private Optional<WithingsDataModel> currentModel;
+
+    public WithingsBridgeHandler(Bridge bridge, AccessTokenInitializableService accessTokenService,
+            HttpClient httpClient) {
+        super(bridge);
+        this.accessTokenService = accessTokenService;
+        this.httpClient = httpClient;
+        this.thingHandlers = new ArrayList<>();
+        this.currentModel = Optional.empty();
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing Withings API bridge handler.");
+
+        configuration = getConfigAs(WithingsBridgeConfiguration.class);
+
+        scheduleTokenInitAndRefresh();
+    }
+
+    public void registerThingHandler(WithingsThingHandler thingHandler) {
+        thingHandlers.add(thingHandler);
+
+        notifyThingHandler(thingHandler);
+    }
+
+    private void connectionSucceed() {
+        updateData();
+        scheduleRefreshJob();
+
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    private void scheduleTokenInitAndRefresh() {
+        tokenRefreshJob = scheduler.scheduleWithFixedDelay(() -> {
+            logger.debug("Initializing API Connection and scheduling token refresh every {}s",
+                    WithingsBindingConstants.TOKEN_REFRESH_SECONDS);
+            try {
+                accessTokenService.init(thing.getUID().getAsString(), configuration);
+
+                boolean isAuthCodeRedeemed = processAuthCodeIfRequired();
+
+                if (!isAuthCodeRedeemed) {
+                    Optional<String> refreshToken = accessTokenService.getRefreshToken();
+                    refreshToken.ifPresent(this::refreshAccessToken);
+                }
+
+                if (accessTokenService.getAccessToken().isPresent()) {
+                    connectionSucceed();
+                } else {
+                    logger.warn("There is no valid access token! Please configure a new auth-code.");
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING,
+                            "Please configure a new auth-code");
+                }
+            } catch (Exception e) {
+                logger.warn("Unable to connect Withings API : {}", e.getMessage(), e);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Withings API access failed, will retry in " + WithingsBindingConstants.TOKEN_REFRESH_SECONDS
+                                + " seconds.");
+            }
+        }, 2, WithingsBindingConstants.TOKEN_REFRESH_SECONDS, TimeUnit.SECONDS);
+    }
+
+    private void scheduleRefreshJob() {
+        refreshJob = scheduler.schedule(() -> {
+            updateData();
+            if (refreshJob != null && !refreshJob.isCancelled()) {
+                refreshJob.cancel(false);
+                refreshJob = null;
+            }
+            scheduleRefreshJob();
+        }, WithingsBindingConstants.REFRESH_SECONDS, TimeUnit.SECONDS);
```

Comment:
Instead of scheduling one-shot tasks, you could use scheduleWithFixedDelay().

Reply:
Oh no, that makes it very difficult to track when something is executed and which caused that. I had negative experiences with it at the Innogy binding and resolved all these usages at the end to make it more maintainable (there were also concurrent tasks which caused problems in combination...).

Reply:
Synchronization between threads should not be achieved (in fact cannot) by above code. Maybe you have solved a synchronization symptom by coincidence but not solved the root cause?

Above code is an anti-pattern, because it's prone to survive `dispose()`: Assume `dispose()` is invoked while the scheduled code runs. Then, the task is cancelled, but the scheduled code will anyhow run until the end and eventually re-schedule the task after the thing is disposed.